### PR TITLE
Overheating quill proposal

### PR DESCRIPTION
**Pull request by amnykon**
Adds heat to quill like stardust. Roughly gives Quill a second of free work over the first 3 seconds of working, while reducing hp slightly to compensate. This ability cools down after 6 seconds. This makes quill slightly better at smaller tasks that require traveling for at-least 6 seconds.

A mod with this change has been uploaded. Try it out: !game Over heating Quill v1

### CODE REVIEW

**Comment by sprunk** [comment]
> makes quill slightly better at smaller tasks that require traveling for at-least 6 seconds

why?

File: scripts/hovercon.lua [hunk 1]
```diff
@@ -17,6 +18,43 @@ local wobble_dir = true
 
 local SIG_BUILD = 1
 
+local function SetupHeat()
+	local ud = UnitDefs[unitDefID]
+	local cp = ud.customParams
+	return tonumber(cp.heat_per_shot), (tonumber(cp.heat_decay) or 0)/30, tonumber(cp.heat_max_slow), tonumber(cp.heat_initial) or 0
+end
+local heatPerShot, heatDecay, heatMaxSlow, currentHeat = SetupHeat()
```
Comment (sprunk) [inline scripts/hovercon.lua:26]: Why a function?
```suggestion
local cp = UnitDefs[unitDefID].customParams
local heatPerShot, heatDecay, heatMaxSlow, currentHeat = tonumber(cp.heat_per_shot), (tonumber(cp.heat_decay) or 0)/30, tonumber(cp.heat_max_slow), tonumber(cp.heat_initial) or 0
```

**Review by sprunk — CHANGES_REQUESTED** [review]

File: units/hovercon.lua [hunk 2]
```diff
@@ -34,7 +38,7 @@ return { hovercon = {
   footprintX          = 2,
   footprintZ          = 2,
   iconType            = [[builder]],
-  maxDamage           = 960,
+  maxDamage           = 940,
```
[the remark this answers is not in the document]
Comment (amnykon) [reply units/hovercon.lua:41]: It is a guess to maintain balancing. I boost in one area requires a nerf in another to maintain balance, though I did not look at all the implications of loosing 20 hp.

**Review by sprunk — COMMENTED** [review]
Alright, was just checking that the health change was conscious and not added by mistake.

**Comment by amnykon** [comment]
> > > > makes quill slightly better at smaller tasks that require traveling for at-least 6 seconds
> > > 
> > > 
> > > why?
> > 
> > 
> > Quill will cool down when traveling, so It will start at a colder temperature. If it does not travel, and starts a new job immediately, it will not cool down and receive a benefit of this ability; however, it will have a higher overall utilization, as it will not take the break to travel. After the 6 seconds, the Quill is cooled down, so farther delay is just wasted.
> 
> Yes, sorry. I understand why the outcome of these mechanics would result in becoming slightly better at tasks with 6s of travel inbetween. The question is "why is making Quill better at spaced tasks a desireable change?".

It is more about adding a unique characteristic. It a simple mechanism that adds uniqueness. Zero-k values the uniqueness of its units. Should this allow apply to constructors? Yes Quill can go on water, but other then that it only has small stat differences to other cons. Amph con can go under water and if being immune to more then half of the weapons is not defensive enough, it also has armor when idle. Heat is a simple mechanism that does not really change its use and can be completely ignored by new players. Yes it makes it stronger in some areas over others, but that is the point.
> 
> Let's see what GF thinks but my 2c:
> 
> * I'm not sure if it would be good to put the (currently rare) mechanic with all its overhead (incl. visual clutter due to heat healthbar, mental clutter because people need to remember this exists, extra logic in places like the space+click box to make sure the unit still displays at 5bp and not 10bp because else it's confusing, etc) for an ability whose main goal seems to boil down to a handful % overall increase in overall mex capping speed.

I think having 10bp there is fine in the details of space+click, as long as the overheating specs are there. Stardust heat should have taken care of that, so it should already be included; however, heat per shot may need to be renamed.

> * instaplopping LLTs gets a comparatively more significant buff which I think (given the recent change to LLT unpack time) would be undesirable.

Assuming quills are cooed down, It would allow a Quill to get a LLT up in 17 seconds instead of 18, 2 quills to get an llt up in 7 seconds instead of 9 seconds, 3 quills to get an LLT up in 3 second instead of 6 seconds. So yes 4 quills approach instant ploping of an llt given that there are sufficient resources. The second LLT will be take longer. So 4 quills and a LLT (590 metal) vs 9 Glaives (585). The 9 Glaives easily kill the LLTs with max of 1 casualty, then finish the 4 quills (2 Glaives each), especially if the quills hp are lowered.

> * isn't it a bit backwards? Hovercraft are about nimbly gliding between places. Maybe the thematic change would be to make the travel faster instead of what happens between the travels? But then note that Quill's "special" ability, in addition to being able to move over water, is already being the fastest non-flying constructor by a significant amount (87 elmo/s, with Mariner and Mason having 75 and 72 respectively).

To me theme of Hovercraft is Higher speed (dagger, bola, halberd, and even scalpel could all be considered a raid unit), high initial damage with long reload(dagger, scalpel, lance), low hp. Now looking at the stats, quills hp is just below mason. That actually surprised me. Quills have more hp than I expect for being hover. I don't think directly matching the theme is good (high instant bp with long reload time). I feel that generating heat fits in with the high initial damage theme.

> * does hovercraft as a whole need it?

No it doesn't need it. My main argument is to add uniqueness. It is a proposal, so it's a discussing point. I learned more about the code base by making this. It was somewhat easy, since I mainly copied what stardust had, and I had help finding the current BP of the quill.  I would be fine if it's rejected; however, I do like the quill.

**Comment by sprunk** [comment]
Given GF's opinion in http://zero-k.info/Forum/Post/248257#248257 the patch looks closeable.

**Pull request closed without merging**